src/sentry/models/organizationslugreservation.py

```diff
@@ -0,0 +1,82 @@
+from __future__ import annotations
+
+from enum import IntEnum
+from typing import Any, Collection
+
+from django.db import models
+from django.utils import timezone
+
+from sentry.backup.scopes import RelocationScope
+from sentry.db.models import BoundedBigIntegerField
+from sentry.db.models.base import control_silo_only_model, sane_repr
+from sentry.db.models.outboxes import ReplicatedControlModel
+from sentry.models import OutboxCategory
+from sentry.services.hybrid_cloud import REGION_NAME_LENGTH
+
+
+class OrganizationSlugReservationType(IntEnum):
+    PRIMARY = 0
+    VANITY_ALIAS = 1
+    TEMPORARY_RENAME_ALIAS = 2
+
+    @classmethod
+    def as_choices(cls):
+        return [(i.value, i.value) for i in cls]
+
+
+@control_silo_only_model
+class OrganizationSlugReservation(ReplicatedControlModel):
+    __relocation_scope__ = RelocationScope.Organization
+    category = OutboxCategory.ORGANIZATION_SLUG_RESERVATION_UPDATE
+    replication_version = 2
+
+    slug = models.SlugField(unique=True, null=False)
+    organization_id = BoundedBigIntegerField(db_index=True, null=False)
+    user_id = BoundedBigIntegerField(db_index=True, null=False)
+    region_name = models.CharField(max_length=REGION_NAME_LENGTH, null=False)
+    reservation_type = BoundedBigIntegerField(
+        choices=OrganizationSlugReservationType.as_choices(),
+        null=False,
+        default=OrganizationSlugReservationType.PRIMARY.value,
+    )
+    date_added = models.DateTimeField(null=False, default=timezone.now, editable=False)
+
+    class Meta:
+        app_label = "sentry"
+        db_table = "sentry_organizationslugreservation"
+        unique_together = (("organization_id", "reservation_type"),)
+
+    __repr__ = sane_repr("slug", "organization_id")
+
+    def __str__(self):
+        return self.slug
+
+    def save(self, *args: Any, **kwds: Any) -> None:
+        assert kwds.get(
+            "unsafe_write", None
+        ), "Cannot write changes to OrganizationSlugReservation unless they go through a provisioning flow"
```

Comment:
This may be a bit heavy handed, but I need some way to ensure that org slug reservations are not created outside of a provisioning/vanity slug creation context.